src/TreeSupport.cpp

```diff
@@ -650,32 +610,119 @@ void TreeSupport::insertDroppedNode(std::unordered_set<Node*>& nodes_layer, Node
+    const auto it = collision_cache_.find(key);
+    if (it != collision_cache_.end())
     {
-        model_avoidance[radius_sample].push_back(model_collision[radius_sample][0]);
-        for (size_t layer_nr = 1; layer_nr < storage.support.supportLayers.size(); layer_nr++)
-        {
-            Polygons previous_layer = model_avoidance[radius_sample][layer_nr - 1].offset(-maximum_move_distance).smooth(5); //Inset previous layer with maximum_move_distance to allow some movement. Smooth to avoid micrometre-segments.
-            previous_layer = previous_layer.unionPolygons(model_collision[radius_sample][layer_nr]);
-            model_avoidance[radius_sample].push_back(previous_layer);
-        }
-#pragma omp atomic
-        completed++;
-#pragma omp critical (progress)
-        {
-            Progress::messageProgress(Progress::Stage::SUPPORT, ((model_collision.size() / 2) + (completed / 2)) * PROGRESS_WEIGHT_COLLISION, model_avoidance.size() * PROGRESS_WEIGHT_COLLISION + storage.support.supportLayers.size() * PROGRESS_WEIGHT_DROPDOWN + storage.support.supportLayers.size() * PROGRESS_WEIGHT_AREAS);
-        }
+        return it->second;
+    }
+    else
+    {
+        return calculateCollision(key);
     }
 }
 
+const Polygons& ModelVolumes::avoidance(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = avoidance_cache_.find(key);
+    if (it != avoidance_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateAvoidance(key);
+    }
+}
+
+const Polygons& ModelVolumes::internal_model(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = internal_model_cache_.find(key);
+    if (it != internal_model_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateInternalModel(key);
+    }
+}
+coord_t ModelVolumes::roundRadius(coord_t radius) const
+{
+    const auto remainder = radius % radius_sample_resolution_;
+    const auto delta = remainder != 0 ? radius_sample_resolution_- remainder : 0;
+    return radius + delta;
+}
+
+const Polygons& ModelVolumes::calculateCollision(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
+
+    auto collision_areas = machine_border_;
+    if (layer < static_cast<int>(layer_outlines_.size())) {
+        collision_areas = collision_areas.unionPolygons(layer_outlines_[layer]);
+    }
+    collision_areas = collision_areas.offset(xy_distance_ + radius, ClipperLib::JoinType::jtRound);
+    const auto ret = collision_cache_.insert({key, std::move(collision_areas)});
+    assert(ret.second);
+    return ret.first->second;
+}
+
+const Polygons& ModelVolumes::calculateAvoidance(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
```

Comment:
Please be more explicit in your variable names.
```suggestion
    const LayerIndex& layer_idx = key.second;
```

Reply:
Variable name clarified. I've left auto for now since the variable name says enough IMO